### ISSUE TITLE
Combine two metrics that have different inputs

### ISSUE THREAD

**Issue by aleSuglia** [PyTorchLightning/metrics#254]
I would like to create a `CompositionalMetric` that is the average of two `Metric`s that have different *inputs*. Currently, seems to me that two metrics can be composed if they have the same positional parameters. The other way would be to use the keyword argument parameters but if you're using two `Accuracy` they would be the same. Can we somehow better redirect the parameters?

**Comment by justusschock** [comment]
Hi @aleSuglia 

What do you propose for that case? Since currently I cannot think of another way to do so.

**Comment by aleSuglia** [comment]
The problem is that they are both `Accuracy` so using keyword arguments doesn't work either (otherwise you mixup the values). Also, I tried to create my own version of an `AverageMetric`:
```
class AverageMetric(Metric):
    """Composition of two metrics with a specific operator which will be executed upon metrics compute """

    def __init__(
            self,
            metric_a: Union[Metric, int, float, Tensor],
            metric_b: Union[Metric, int, float, Tensor, None]
    ):
        """
        Args:
            operator: the operator taking in one (if metric_b is None)
                or two arguments. Will be applied to outputs of metric_a.compute()
                and (optionally if metric_b is not None) metric_b.compute()
            metric_a: first metric whose compute() result is the first argument of operator
            metric_b: second metric whose compute() result is the second argument of operator.
                For operators taking in only one input, this should be None
        """
        super().__init__()

        if isinstance(metric_a, Tensor):
            self.register_buffer("metric_a", metric_a)
        else:
            self.metric_a = metric_a

        if isinstance(metric_b, Tensor):
            self.register_buffer("metric_b", metric_b)
        else:
            self.metric_b = metric_b

    def _sync_dist(self, dist_sync_fn: Callable = None) -> None:
        # No syncing required here. syncing will be done in metric_a and metric_b
        pass

    def update(self, *args, **kwargs) -> None:
        metric_a_args = kwargs["a"]
        if isinstance(self.metric_a, Metric):
            self.metric_a.update(*metric_a_args)

        metric_b_args = kwargs["b"]
        if isinstance(self.metric_b, Metric):
            self.metric_b.update(*metric_b_args)

    def compute(self) -> Any:

        # also some parsing for kwargs?
        if isinstance(self.metric_a, Metric):
            val_a = self.metric_a.compute()
        else:
            val_a = self.metric_a

        if isinstance(self.metric_b, Metric):
            val_b = self.metric_b.compute()
        else:
            val_b = self.metric_b

        return (val_a + val_b) / 2

    def reset(self) -> None:
        if isinstance(self.metric_a, Metric):
            self.metric_a.reset()

        if isinstance(self.metric_b, Metric):
            self.metric_b.reset()

    def persistent(self, mode: bool = False) -> None:
        if isinstance(self.metric_a, Metric):
            self.metric_a.persistent(mode=mode)
        if isinstance(self.metric_b, Metric):
            self.metric_b.persistent(mode=mode)

    def __repr__(self) -> str:
        _op_metrics = f"(\n  average(\n    {repr(self.metric_a)},\n    {repr(self.metric_b)}\n  )\n)"
        repr_str = self.__class__.__name__ + _op_metrics

        return repr_str
```
Now imagine I have `train_actions_accuracy` and `train_objects_accuracy`, then I create an `overall` metric using the above class. However, I'm having issues in logging it if I do:

```
        self.log("train_overall", self.train_overall, on_epoch=True, prog_bar=True, on_step=True,
                 tbptt_reduce_fx=torch.sum)
        self.log("train_actions_accuracy", self.train_actions_accuracy, on_epoch=True, tbptt_reduce_fx=torch.sum)
        self.log("train_objects_accuracy", self.train_objects_accuracy, on_epoch=True, tbptt_reduce_fx=torch.sum)
```
When running my training I get:
```
you tried to log Accuracy() which is not currently supported. Try a dict or a scalar/tensor.
```

**Comment by justusschock** [comment]
@aleSuglia got it. you should be able to do it like this (I think):

```
acc1 = Accuracy()
acc2 = Accuracy()
mean_acc = (acc1 + acc2)/2

acc1.update(some_args_here)
acc2.update(some_args_here)
self.log(mean_acc)
```

since the compositional metric only takes references to those metrics. 

However, I did not test this. If you run into issues, please report back :)

**Comment by aleSuglia** [comment]
Would this log also `acc1` and `acc2`?

**Comment by justusschock** [comment]
No, this would only log the mean. 

I am not sure what happens when you change this to 
```python
acc1 = Accuracy()
acc2 = Accuracy()
mean_acc = (acc1 + acc2)/2

acc1.update(some_args_here)
acc2.update(some_args_here)
self.log('acc1', acc1)
self.log('acc2', acc2)
self.log('mean_acc', mean_acc)
```

since then `compute` has already been called when you reach the last log statement. 
cc @SkafteNicki : Do you know what happens then/if the behaviour will be correct?

**Comment by aleSuglia** [comment]
Also, you seems that you need to call the update of `mean_acc` as well otherwise it doesn't get updated.

**Comment by SkafteNicki** [comment]
I propose that you subclass accuracy twice and thereby overwrite the signature of `update`:
```python
#%%
from torchmetrics import Accuracy
import torch

class Acc1(Accuracy):
    def update(self, preds_1, target_1):
        super().update(preds_1, target_1)
        
    def compute(self):
        print('acc1 called')
        return super().compute()
        
class Acc2(Accuracy):
    def update(self, preds_2, target_2):
        super().update(preds_2, target_2)

    def compute(self):
        print('acc2 called')
        return super().compute()
        
acc1 = Acc1()
acc2 = Acc2()
mean_acc = (acc1 + acc2) / 2

mean_acc.update(preds_1 = torch.randint(10, (100,)),
                target_1 = torch.randint(10, (100,)),
                preds_2 = torch.randint(10, (100,)),
                target_2 = torch.randint(10, (100,)))

res_mean = mean_acc.compute()
res_acc1 = acc1.compute()
res_acc2 = acc2.compute()
```
note only `update` is required to be overridden, but this example will show that `compute` is only called **once** for both acc1 and acc2 since we cache the computed value such that repeated call to compute just returns the cache value. Here is the relevant code:
https://github.com/PyTorchLightning/metrics/blob/cda5dbd58c6c6ea55d620a8aa521789f4b56b125/torchmetrics/metric.py#L236-L238

As @justusschock mentions the correct way to log would be:
```python
self.log('acc1', acc1)
self.log('acc2', acc2)
self.log('mean_acc', mean_acc)
```
We cannot enforce that compositional metrics should also log the value of the children as this is maybe not something that all users wants.

**Comment by aleSuglia** [comment]
Oh thanks for the neat example @SkafteNicki. What would be the correct way of integrating them in a LightningModule?

**Comment by SkafteNicki** [comment]
Personally I would do it this way:
```python
class MyModule(pl.LightningModule):
    def __init__(self, ...):
        ...
        self.acc1 = Acc1()
        self.acc2 = Acc2()
        self.mean_acc = (acc1 + acc2) / 2

    def  validation_step(self, ...):
        ...
        self.mean_acc.update(preds_1 = preds_1, target_1 = target_1, preds_2 = preds_2, target_2 = target_2)

    def validation_epoch_end(self, ...):
        ...
        res_mean = self.mean_acc.compute()
        res_acc1 = self.acc1.compute()
        res_acc2 = self.acc2.compute()

        self.log('acc1', res_acc1)
        self.log('acc2', res_acc2)
        self.log('mean_acc', res_mean)

        self.mean_acc.reset()
```

**Comment by aleSuglia** [comment]
Sorry for the late response. Should `log()` take care of `reset()` when `on_epoch=True`? Also, why do we have a call to `reset` only for `mean_acc`?

**Comment by SkafteNicki** [comment]
1. Please note that in my example I am not passing the metric object directly, but instead passing the computed tensors to `log` which is the reason that I have to call `reset` manually
2. `reset` of `mean_acc` will call `reset` of the `acc1` and `acc2`

**Comment by XiaomoWu** [comment]
> Hi @aleSuglia
> 
> What do you propose for that case? Since currently I cannot think of another way to do so.


A compositional metric that takes in different inputs is very handy when doing multi-task learning, where `total_loss = w1*loss(y1, t1) + (1-w1)*loss(y2, t2)`

**Comment by SkafteNicki** [comment]
@XiaomoWu as stated in my above commen (https://github.com/PyTorchLightning/metrics/issues/254#issuecomment-842435391), In that case overwriting the loss metric twice should give you exactly that:
```python
class LossMetric1(LossMetric):
    def update(self, y1, t1):
        super().update(y1, t1)
class LossMetric2(LossMetric):
    def update(self, y2, t2):
        super().update(y2, t2)
total_loss = w1*LossMetric1() + (1-w1)*LossMetric2()
val = total_loss(y1=y1, y2=y2, t1=t1, t2=t2)
```
